chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2513 @@
+    "step": STEP_LIST,
+    "step291": STEP291_LIST,
+    "tdrab": TDRAB_LIST,
+    "tdrch": TDRCH_LIST,
+    "timeout": TIMEOUT_LIST,
+    "txled": COLOR_LIST,
+    "txpower": TXPOWER_LIST,
+    "txpower3": TXPOWER3_LIST,
+    "voice": VOICE_LIST,
+    "vox": VOX_LIST,
+    "widenarr": BANDWIDTH_LIST,
+    "workmode": WORKMODE_LIST,
+    "wtled": COLOR_LIST
+}
+
+GMRS_FREQS1 = [462.5625, 462.5875, 462.6125, 462.6375, 462.6625,
+               462.6875, 462.7125]
+GMRS_FREQS2 = [467.5625, 467.5875, 467.6125, 467.6375, 467.6625,
+               467.6875, 467.7125]
+GMRS_FREQS3 = [462.5500, 462.5750, 462.6000, 462.6250, 462.6500,
+               462.6750, 462.7000, 462.7250]
+GMRS_FREQS_B = GMRS_FREQS2 * 3 + GMRS_FREQS3 * 2
+
+GMRS_FREQ_A = GMRS_FREQS1 + GMRS_FREQS3
+
+HAM_FREQS = [162.55000, 162.40000, 162.47500, 162.42500, 162.45000,
+             162.50000, 162.52500, 161.65000, 161.77500, 161.75000]
+HAM_NAME = ["NOAA 1", "NOAA 2", "NOAA 3", "NOAA 4", "NOAA 5", "NOAA 6",
+            "NOAA 7", "NOAA 8", "NOAA 9", "NOAA 10", "NOAA 11"]
+
+ALL_MODEL = ["TD-H8", "TD-H8-HAM", "TD-H8-GMRS"]
+
+def _do_status(radio, block):
+    status = chirp_common.Status()
+    status.msg = "Cloning"
+    status.cur = block
+    status.max = radio._memsize
+    radio.status_fn(status)
+
+TD_H8 = b"\x50\x56\x4F\x4A\x48\x1C\x14"
+
+def _upper_band_from_data(data):
+    return data[0x03:0x04]
+
+def _upper_band_from_image(radio):
+    return _upper_band_from_data(radio.get_mmap())
+
+def _firmware_version_from_data(data, version_start, version_stop):
+    version_tag = data[version_start:version_stop]
+    return version_tag
+
+def _firmware_version_from_image(radio):
+    version = _firmware_version_from_data(radio.get_mmap(),
+                                          radio._fw_ver_file_start,
+                                          radio._fw_ver_file_stop)
+    # LOG.debug("_firmware_version_from_image: " + util.hexprint(version))
+    return version
+
+def _do_ident(radio, magic, secondack=True):
+    serial = radio.pipe
+    serial.timeout = 1
+    
+    LOG.info("Sending Magic: %s" % util.hexprint(magic))
+    serial.write(magic)
+    ack = serial.read(1)
+    
+    # for byte in magic:
+    #     serial.write(bytes([byte]))
+    #     time.sleep(0.01)
+    # ack = serial.read(1)
+
+    if ack != b"\x06":
+        if ack:
+            pass
+            # LOG.debug(repr(ack))
+        raise errors.RadioError("Radio did not respond")
+
+    serial.write(b"\x02")
+
+    response = b""
+    for i in range(1, 9):
+        byte = serial.read(1)
+        response += byte
+        if byte == b"\xDD":
+            break
+        
+    if len(response) in [8, 12]:
+        # DEBUG
+        LOG.info("Valid response, got this:")
+        LOG.debug(util.hexprint(response))
+        
+        # Determine the walkie-talkie mode
+        # TDH8 have three mode:ham, gmrs and normal
+        global MODE_RADIO 
+        MODE_RADIO = util.hexprint(response)[31:37]
```

Comment:
Right here, you know which radio the user has selected (because `radio`) and you know what radio they have connected (because `response`). Compare the selected model with the response you get and either fail right here, or continue depending on whether they match. Don't wait until the end of `sync_in()` to check/fail.

Reply:
Yes, I will use bcd, but according to the walkie-talkie protocol I got, I still need to use some characters in the codec to judge which type of tmode it is. If I use integers, forgive me for being an idiot Rookie, it will be more troublesome to deal with.

Reply:
I've modified judgment before doing the cloning routine and he now works as you said.

Reply:
> Yes, I will use bcd, but according to the walkie-talkie protocol I got, I still need to use some characters in the codec to judge which type of tmode it is. If I use integers, forgive me for being an Rookie, it will be more troublesome to deal with.

I think it will be the same, but I can't really say without seeing the details. So, if you can update this to use BCD and show what is harder or more work, perhaps either I will understand, or I can help make it work.

All the other drivers with this kind of data format use bcd, so I think we should try very hard to keep this consistent.

Reply:
Let me give you an example. If the tone is 67.0, it is [0x70, 0x06] according to the TDH8 protocol, so I need to convert the above to 67.0, execute the clone process get_memory(), and I need to convert lin2 to a string "0670 ", and then converted to int type for display.
For downloading, I finally need to process the format as [0x70,0x06].

If DTCS or RX DTCS is 023N, it is [0x23, 0x80] according to the protocol, and 8 is the identification of N. If DTCS or RX DTCS is 023R, it is [0x23, 0xC0] according to the protocol, and C is the identification of R. To execute the cloning process, I need to convert it to "C023", judge it as R according to "C", and then use slices to convert 023 to int type for use.

So it is more convenient to use types like `u8` [0x00, 0x00].

Reply:
Thanks for the example. What you describe is very similar to (if not exactly the same as) many other drivers. Here is an example of another driver that behaves almost exactly as you describe:

https://github.com/kk7ds/chirp/blob/master/chirp/drivers/tk8180.py#L633-L650

That one uses a `ul16` but an `lbcd` will be equivalent for you in terms of the behavior when you call `int()` on it. Many (many) others do the same, but not all of them are quite as well organized as that one, so it's a good example. Here's an example of a driver that uses an `lbcd` tone value very similar to yours, maybe even identical:

https://github.com/kk7ds/chirp/blob/master/chirp/drivers/h777.py#L350-L359

So yes, please do use a real type and not individual bytes with a string in the middle for this.